### ISSUE TITLE
New option to unsubscribe all websocket connections for a username

### ISSUE THREAD

**Issue by fender** [chrisdew/jsonspace#33]
@chrisdew We're running into an edge case with unsubscribe. For example, if I am on my iPhone and iPad subscribed to the same channel and on my iPhone I say I want to leave that channel (i.e. unsubscribe) then it should also unsubscribe the iPad too. However, if I closed the app on my iPhone (because I wanted to continue just on my iPad) it also sends an unsubscribe message but we don't want to unsubscribe the iPad.

So, in short, can we add an option to the unsubscribe method to also unsubscribe all connections for the same channel with the same username? That should also issue the `left` message to all watchers/subscribers as if it were the last unsubscribe.

**Comment by chrisdew** [comment]
> However, if I closed the app on my iPhone (because I wanted to continue just on my iPad) it also sends an unsubscribe message but we don't want to unsubscribe the iPad.

"unsubscribe" just unsubscribes one username/channel/websocket combination and so should not unsubscribe the iPad when you unsubscribe from the iphone.  If you are seeing this, it is a bug.

> So, in short, can we add an option to the unsubscribe method to also unsubscribe all connections for the same channel with the same username? That should also issue the left message to all watchers/subscribers as if it were the last unsubscribe.

Yes, I can do this, but I want to clarify whether you are seeing a bug, first.

**Comment by fender** [comment]
No, we're not seeing a bug with the server at the moment. Just need to have a new feature to force unsubscribe even if it isn't the last connection.

**Comment by chrisdew** [comment]
This is turning out to be more difficult than I had guessed.
The generation of "left" messages assumes that unsubscribes are done one at a time, which is now no longer true.

**Comment by chrisdew** [comment]
Only the last connection will get a left message, as it's the only connection still subscribed when the the number of subscriptions goes from 1->0.

**Comment by chrisdew** [comment]
I cannot keep the client in sync without re-introducing 'unsubscribed' messages.
Note: I understand that you want to limit bandwidth, but I cannot see how you can have a robust client without "subscribed", "unsubscribed" and "published" messages being sent to clients.

I am going to reintroduce the "subscribed" and "unsubscribed" message.  You can ignore them, but they are logically required for correct client operation.

**Comment by chrisdew** [comment]
Do you want this feature, at the cost of reintroducing "subscribed" and "unsubscribed" message to the client?

(I thought I'd ask, as its an hours work, which may be wasted, if you're not prepared to accept these client acks.)

**Comment by fender** [comment]
@chrisdew If we don't have to change any existing logic, that's OK. However, if it does mess with how things worked until now, can we try to find another solution? I've spent quite some time nailing down the client-side logic. In case we can come up with an easy solution, here is the specific use case:

On the app, users click a Join button to join a chat room (i.e. subscribe). If they have other devices or the website open and watching the same channel (happens automatically when viewing a tournament) then they receive a "joined" message which causes them to also subscribe. All devices are in sync, so far so good.

Now, if the user decides they want to leave the chat room (i.e. unsubscribe) they would click a "Leave" button. However, as there are multiple devices that are all subscribed, a "left" message doesn't get broadcasted because it wasn't the last unsubscribe. A single device unsubscribed but the others are still subscribed. That's fine if the user just turned off the app or disconnected on one of the devices, but they specifically said they want to leave. So we need to handle this case.

Perhaps a new method like "exit"? A user could send that and it just sends a '"left" immediately?

**Realization**
To be honest, I could do this without needing any server-side changes. I could publish a message with the data set as "__EXIT__" or something have clients listen to that and unsubscribe when they read it. It's a little bit dirty, but I have already done that for something specific to the app business logic.

Thoughts?

**Comment by chrisdew** [comment]
Currently the clients *assume* that what they've requested of the server has happened.

I don't like this, but I accept and understand your reasons.

If you don't want to turn things around, so that the clients listen to the server to find out whether their subscribe and unsubscribe *requests* have been successful, then **I think you should go with your realization**.  

(In order for one connection to unsubscribe another, the clients would have to start listening to the server for new "unsubscribe" messages, rather than assuming that they always successfully instruct the server.)

I expect that you will find another case which will require this turn-around, in a few weeks. 

I hadn't realised that there was this back-channel which caused other devices to join without action from the device's user.  I think this is the cause of the tension between subscription-state managed by the pubsub server and subscription-state managed by the rest of the website.

I'll close this issue for now.

**Comment by fender** [comment]
Makes sense. We can revisit if I start to see some issues. I totally
understand your original logic.
